src/codegen/compile.ml

```diff
@@ -8626,21 +8193,222 @@ module Stabilization = struct
+  If logical size N > 0:
+    [0..4]          0 (first word is backed up at `end-8`)
+    [4..N]          <stable memory>
+            <zero padding>
+    [end-16..end-8] <size N>
+    [end-8..end-4]  <first word>
+    [end-4..end]    <new version>
+  ending at page boundary
+  Note: The first word must be empty to distinguish this version from the Candid legacy version 0 (which has first word != 0).
+*)
+module NewStableMemory = struct
+  let physical_size env =
+    IC.system_call env "stable64_size" ^^
+    compile_shl64_const (Int64.of_int page_size_bits)
+
+  let store_at_end env offset typ get_value =
+    physical_size env ^^
+    compile_sub64_const offset ^^
+    get_value ^^
+    match typ with
+    | I32Type -> StableMem.write_word32 env
+    | I64Type -> StableMem.write_word64 env
+    | _ -> assert false
 
-    let get_static_roots = E.add_fun env "get_static_roots" (Func.of_body env [] [I32Type] (fun env ->
-      compile_unboxed_const static_roots
-    )) in
+  let read_from_end env offset typ =
+    physical_size env ^^
+    compile_sub64_const offset ^^
+    match typ with
+    | I32Type -> StableMem.read_word32 env
+    | I64Type -> StableMem.read_word64 env
+    | _ -> assert false
 
-    E.add_export env (nr {
-      name = Lib.Utf8.decode "get_static_roots";
-      edesc = nr (FuncExport (nr get_static_roots))
-    })
+  let clear_at_end env offset typ =
+    store_at_end env offset typ 
+    (match typ with
+    | I32Type -> compile_unboxed_const 0l
+    | I64Type -> compile_const_64 0L
+    | _ -> assert false
+    )
+
+  let logical_size_offset = 16L
+  let first_word_backup_offset = 8L
+  let version_offset = 4L
+
+  let upgrade_version env =
+    StableMem.set_version env ^^
+    StableMem.get_version env ^^
+    compile_eq_const StableMem.legacy_version_no_stable_memory ^^
+    StableMem.get_version env ^^
+    compile_eq_const StableMem.legacy_version_some_stable_memory ^^
+    G.i (Binary (Wasm.Values.I32 I32Op.Or)) ^^
+    (G.if0
+      (compile_unboxed_const StableMem.version_stable_heap_no_regions ^^
+      StableMem.set_version env)
+      G.nop) ^^
+    StableMem.get_version env ^^
+    compile_eq_const StableMem.legacy_version_regions ^^
+    (G.if0
+      (compile_unboxed_const StableMem.version_stable_heap_regions ^^
+      StableMem.set_version env)
+      G.nop)
+      
+  let grow_size env amount =
+    StableMem.get_mem_size env ^^
+    compile_shl64_const (Int64.of_int page_size_bits) ^^
+    compile_const_64 amount ^^
+    StableMem.ensure env
+
+  let backup env =
+    let (set_first_word, get_first_word) = new_local env "first_word" in
+    physical_size env ^^
+    G.i (Test (Wasm.Values.I64 I64Op.Eqz)) ^^
+    G.if0
+      G.nop
+      begin
+        (* read and clear first word *)
+        compile_const_64 0L ^^ StableMem.read_word32 env ^^ set_first_word ^^
+        compile_const_64 0L ^^ compile_unboxed_const 0l ^^ StableMem.write_word32 env ^^
+
+        grow_size env logical_size_offset ^^
```

Comment:
I'm not convinced this grow_size is ensuring what you want to justify the following `store_at_end` calls.
Should it maybe be
```suggestion
        grow_size env 65536L ^^
```
To ensure an additional page is physically allocated?

Reply:
My thought was that I do not necessarily need a new page, just at least 16 bytes of free space. If the current page has at least that free capacity, the data would be written at the end of that page (`store_at_end`). Otherwise, an additional page would be allocated and I would write at that end.
Maybe I am wrong or that's a too weak assumption?